consensus/src/threshold_simplex/types.rs

```diff
@@ -64,45 +54,284 @@ pub trait Attributable {
+    pub signer: u32,
+    /// Scheme-specific signature or share produced for the vote context.
+    pub signature: S::Signature,
+}
+
+impl<S: SigningScheme> PartialEq for Vote<S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.signer == other.signer && self.signature == other.signature
+    }
+}
+
+impl<S: SigningScheme> Hash for Vote<S> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.signer.hash(state);
+        self.signature.hash(state);
+    }
+}
+
+impl<S: SigningScheme> Write for Vote<S> {
+    fn write(&self, writer: &mut impl BufMut) {
+        self.signer.write(writer);
+        self.signature.write(writer);
+    }
+}
+
+impl<S: SigningScheme> EncodeSize for Vote<S> {
+    fn encode_size(&self) -> usize {
+        self.signer.encode_size() + self.signature.encode_size()
+    }
+}
+
+impl<S: SigningScheme> Read for Vote<S> {
+    type Cfg = ();
+
+    fn read_cfg(reader: &mut impl Buf, _: &()) -> Result<Self, Error> {
+        let signer = u32::read(reader)?;
+        let signature = S::Signature::read(reader)?;
+
+        Ok(Self { signer, signature })
+    }
+}
+
+/// Result of verifying a batch of votes.
+pub struct VoteVerification<S: SigningScheme> {
+    /// Contains the votes accepted by the scheme.
+    pub verified: Vec<Vote<S>>,
+    /// Identifies the participant indices rejected during batch verification.
+    pub invalid_signers: Vec<u32>,
+}
+
+impl<S: SigningScheme> VoteVerification<S> {
+    /// Creates a new `VoteVerification` result.
+    pub fn new(verified: Vec<Vote<S>>, invalid_signers: Vec<u32>) -> Self {
+        Self {
+            verified,
+            invalid_signers,
+        }
+    }
 }
 
-// Constants for domain separation in signature verification
-// These are used to prevent cross-protocol attacks and message-type confusion
-pub const SEED_SUFFIX: &[u8] = b"_SEED";
-pub const NOTARIZE_SUFFIX: &[u8] = b"_NOTARIZE";
-pub const NULLIFY_SUFFIX: &[u8] = b"_NULLIFY";
-pub const FINALIZE_SUFFIX: &[u8] = b"_FINALIZE";
+/// Cryptographic surface required by `threshold_simplex`.
+///
+/// A `SigningScheme` produces validator votes, validates them (individually or in
+/// batches), assembles quorum certificates, checks recovered certificates and, when
+/// available, derives randomness for leader rotation. Implementations may override the
+/// provided defaults to take advantage of scheme-specific batching strategies.
+pub trait SigningScheme: Clone + Debug + Send + Sync + 'static {
+    type Signature: Clone
+        + Debug
+        + PartialEq
+        + Eq
+        + Hash
+        + Send
+        + Sync
+        + EncodeSize
+        + Write
+        + Read<Cfg = ()>;
+
+    type Certificate: Clone
+        + Debug
+        + PartialEq
+        + Eq
+        + Hash
+        + Send
+        + Sync
+        + EncodeSize
+        + Write
+        + Read<Cfg = Self::CertificateCfg>;
+
+    type Randomness: Clone + EncodeSize + Write + Send;
```

Comment:
Use `Encode` instead of `EncodeSize + Write`. You might even be able to use `EncodeFixed` if we always expect the randomness to be encodable into a fixed-size array